WIP: Various changes

bgamari opened the pull request:
*While preparing this PR I realized that the history really needs some love. I'll try to do this today*

This is the result of a weekend of hacking on my part to try to get Ward into a state where it can be used to verify GHC's runtime system:

 * Compatibility with recent GHCs
 * Avoid trivial uses of `syb`'s `everything` combinator as it is extremely expensive
 * Hide representation of  `CallMap` behind a `newtype` and drop custom serialiser in favor of JSON. JSON is a bit harder to read but a bit faster and is significantly easier implementation-wise
 * Deepseq `language-c` ASTs during graph preparation
 * Use a proper ADT to represent `Enforcement` instead of `These`. Afterall, it's possible that we might want to extend the enforcement specifier language in the future
 * Use strict maps wherever possible
 * Intern `CallMap` `Ident`s to reduce memory consumption due to duplication
 * Refactor pipeline to allow parallel preparation of `CallMap`s from source files
 * A rebase of #17

lambdageek committed:
[Check.Permissions] Pull one more arg out of propagatePermissionsNode

lambdageek committed:
[Check.Permissions] Use the Any monoid for growing (NFC)

Maybe this reads a little better?

lambdageek committed:
[Check.Permissions] Pull processCallTree  out of propagatePermissionsNode

lambdageek committed:
Refactor CallTree into sequence of choices

Basically the canonical form we want is a sequence of calls and choices
where the branches of the choice are themselves sequences.

lambdageek committed:
Change CallSequence to use a Data.Sequence.Seq instead of a list

lambdageek committed:
[Check.Permissions] factor out permissionsPresenceFromCalleeActions

lambdageek committed:
[Check.Permissions] Refactor permissionsPresenceFromCalleeActions

lambdageek committed:
Reformat some comments to make Haddock happy (NFC)

lambdageek committed:
Add Algebra.Algebra - (bounded) join & meet semilattice typeclasses

This is cribbed from
https://hackage.haskell.org/package/lattices/docs/Algebra-Lattice.html

except my HashMap instance uses unionWith instead of union

lambdageek committed:
Use a lattice representation of PermissionPresenceSet

Represent PermissionPresence as a product lattice of Usage and Capability.

* Usage is a simple two element lattice - either we know nothing about how a
  permission name is used, or we know that it's used by a call that has a 'Uses'
  action.

* Capability is a 4 element lattice that tracks what we know about a permission
  name at each site in a call sequence: CapUnknown (bottom) - we don't know anything,
  CapHas - the site has the capability, CapLacks - the site lacks
  the capablity, CapConflict - we learned conflicting information (both
  'CapHas' and 'CapLacks') along different paths in the call tree.

* PermissionPresence is a simple product lattice: its bottom is the pair bottom
  bottom and its join and partial order are given component-wise.

The PermissionPresence lattice extends to a lattice on
PermissionPresenceSet (now a finite map from PermissionName to
PermissionPresence): the empty set is the bottom element, and the join of two
finite maps takes their union applying the join operation of PermissionPresence
when the same name is present in both maps.

With this, we can see that most of the propagation during inference is simple
joins of old and new information except when propagating forward through 'Call'
site which has a 'Grant' or 'Revoke' action which overwrites the Capability
portion of the prior information (which are in some sense stale from the
previous iteration).

lambdageek committed:
[test] fixup golden strings

lambdageek committed:
Refactor: don't compute PermissionAction -> PermissionPresence multiple ways

both the initial construction of PermissionPresence from PermissionAction, and
the computation of pre- and post-conditions when working at a 'Call' site can
be decomposed in the same way

lambdageek committed:
nodeAnnotations don't need to be mutable

bgamari committed:
Ensure that declarations don't override definition's calls

bgamari committed:
Ensure that output via CompilerOutput occurs on separate lines

bgamari committed:
Don't use everything

bgamari committed:
Make CallMap a newtype

bgamari committed:
Drop some redundant imports

bgamari committed:
Drop redundant everywhere

bgamari committed:
Fix unused binders

bgamari committed:
Replace callmap format with JSON

bgamari committed:
Fix some warnings

bgamari committed:
Show filename in error

bgamari committed:
Simplify call map dumping

bgamari committed:
Deepseq language-c ASTs

bgamari committed:
Make Enforcement its own ADT

bgamari committed:
Intern CallMaps and concurrently read

bgamari committed:
Use strict maps

bgamari committed:
Refactor parallelism

Build CallMaps from the individual inputs and smash them together at the
end.

bgamari committed:
Concurrent exception safety

bgamari committed:
Fix Monoid instance for CallMap

bgamari committed:
Add Show instance for Function

bgamari committed:
Add useful Show instance for CallMap

bgamari committed:
Graph: Prefix statics with source filename

evincarofautumn commented on src/Graph.hs:
```diff
@@ -382,9 +382,9 @@ callMapFromNameMap = CallMap . Map.mapWithKey fromEntry
+  CCall (CVar (Ident actionName _ _) _) permissions _pos <- select expressions
   permissionSpec <- select permissions
-  (permission, mSubject) <- case permissionSpec of
+  (permission, _mSubject) <- case permissionSpec of
```
I think I left this unmarked with `_` because it was supposed to be the “hook” where we start to handle #37, albeit with `CVar` instead of `CConst`

evincarofautumn commented on src/DumpCallMap.hs:
```diff
@@ -1,157 +1,22 @@
-
--- | Encode the given callmap as a "Data.ByteString
-encodeCallMap :: CallMap -> B.Builder
-encodeCallMap (CallMap cm) = form "callmap" True $ Map.foldMapWithKey encodeCallMapItem cm
```
*In principle* from a release management perspective I like to retain support for old formats, but honestly I don’t think anyone is really depending on this, and a user can just blow away and regenerate the files. So ±0 on that.

bgamari replied:
While I can understand that I think the benefits of this format are *far* outweighed by the persistent maintenance cost that it incurs (not to mention the fact that it showed up pretty heavily in my profiling, so there was clearly some performance on the table).

evincarofautumn replied:
Agreed, just getting this down for posterity.

evincarofautumn commented on src/Types.hs:
```diff
@@ -243,7 +256,7 @@ type CallMapParseError = ParsecErr.ParseError
 data ProcessingUnitParseError =
   CSourceUnitParseError CParser.ParseError
-  | CallMapUnitParseError !CallMapParseError
+  | CallMapUnitParseError String
```
Is `CallMapParseError` now unused? We could change it:

```diff
-type CallMapParseError = ParsecErr.ParseError
+type CallMapParseError = String
```

bgamari replied:
It seemed a bit unnecessary as the data constructor is already pretty descriptive but I have gone ahead and reintroduced the type synonym.

bgamari committed:
cabal: Fix ordering of build-depends

bgamari committed:
Add semigroups dependency

For compatibility with older GHCs.

bgamari committed:
Bump language-c lower bound

We need Generic instances for the AST types.

bgamari committed:
Generate Travis configuration with haskell-ci

bgamari committed:
Move most sources to library component

This allows us to avoid compiling the project once for the executable
and again for the testsuite. Moreover, we can avoid duplicate the
build-depends.

bgamari committed:
cabal: Add test files to extra-source-files

bgamari committed:
gitignore: Add dist-newstyle

bgamari committed:
Silence some `cabal check` issues

bgamari committed:
Fix compatibilty with GHC 8.2

bgamari commented:
It seems that 8.2.2 hangs during CI. I've removed this configuration from the Travis matrix.

bgamari commented:
Any objection to merging this, @evincarofautumn?

evincarofautumn reviewed (approved):
LGTM and thanks for your help! Merging
